lib/pages/search/public_room/search_public_room_controller.dart

```diff
@@ -0,0 +1,186 @@
+    searchResultsNotifier.value = PresentationSearchPublicRoom(
+      searchResults: [],
+    );
+  }
+
+  void _searchPublicRoom() {
+    _publicRoomInteractor
+        .execute(
+          filter: _filter,
+          limit: _limitPublicRoomSearchFilter,
+          server: getServerName(_filter?.genericSearchTerm),
+        )
+        .listen(
+          (searchResult) => _handleListenSearchPublicRoom(searchResult),
+        );
+  }
+
+  void _handleListenSearchPublicRoom(Either<Failure, Success> searchResult) {
+    searchResult.fold((failure) => _resetSearchResults(), (success) {
+      if (!searchTermIsNotEmpty) {
+        return;
+      }
+
+      if (success is PublicRoomSuccess) {
+        if (success.publicRoomsChunk == null) {
+          searchResultsNotifier.value = PresentationSearchPublicRoomEmpty();
+        } else if (success.publicRoomsChunk!.isEmpty) {
+          searchResultsNotifier.value = PresentationSearchPublicRoomEmpty();
+        } else {
+          searchResultsNotifier.value = PresentationSearchPublicRoom(
+            searchResults: success.publicRoomsChunk!,
+          );
+        }
+      }
+    });
+  }
+
+  void _handleKeywordIsRoomId() {
+    searchResultsNotifier.value = PresentationSearchPublicRoomEmpty();
+  }
+
+  void _viewRoom(BuildContext context, String roomId) {
+    context.go('/rooms/$roomId');
+  }
+
+  String? getServerName(String? roomIdOrAlias) {
+    if (roomIdOrAlias != null) {
+      return roomIdOrAlias.getServerNameFromRoomIdOrAlias();
+    }
+    return null;
+  }
+
+  void joinRoom(
+    BuildContext context,
+    String roomIdOrAlias,
+    String? server,
+  ) async {
+    final client = Matrix.of(context).client;
+    final result = await TwakeDialog.showFutureLoadingDialogFullScreen(
+      future: () => client.joinRoom(
+        roomIdOrAlias,
+        serverName: server != null ? [server] : null,
+      ),
+    );
+    if (result.error == null) {
+      if (client.getRoomById(result.result!) == null) {
+        await client.onSync.stream.firstWhere(
+          (sync) => sync.rooms?.join?.containsKey(result.result) ?? false,
+        );
+      }
+      if (!client.getRoomById(result.result!)!.isSpace) {
+        context.go('/rooms/${result.result!}');
+      }
+      return;
+    }
+  }
+
+  void handlePublicRoomActions(
+    BuildContext context,
+    PublicRoomsChunk room,
+    PublicRoomActions action,
+  ) {
+    switch (action) {
+      case PublicRoomActions.join:
+        joinRoom(
+          context,
+          room.roomId,
+          getServerName(room.roomId),
+        );
+        break;
+      case PublicRoomActions.view:
+        _viewRoom(context, room.roomId);
+        break;
+    }
+  }
+
+  PublicRoomActions? getAction(
+    BuildContext context,
+    PublicRoomsChunk room,
+  ) {
+    final client = Matrix.of(context).client;
+    if (client.getRoomById(room.roomId) != null) {
+      return PublicRoomActions.view;
+    } else if (room.joinRule == 'public') {
```

Comment:
should create an enum

Reply:
I think we will create it later when implement other rules